### ISSUE TITLE
rclone dedupe fails to remove duplicates (Drive)

### ISSUE THREAD

**Issue by tcf909** [rclone/rclone#4321]
#### What is the problem you are having with rclone?
When running rclone dedupe against a remote the files are not actually deduplicated.

Running the same rclone dedupe command multiple times results in the same result (finding dupe files and reporting they are being removed).

I've tried on the raw remote (Drive) as well as the encrypted remote (Drive > Crypt). Same result.

#### What is your rclone version (output from `rclone version`)
```
rclone v1.52.0
- os/arch: linux/amd64
- go version: go1.14.3
```

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)
Linux Debian 10.x

####  Which cloud storage system are you using? (eg Google Drive)
Google Drive

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)
```
rclone dedupe -v --fast-list --checkers=24 --dedupe-mode skip Remote:
```

#### A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)
 Lots of:
```
2020/06/08 19:26:07 NOTICE: Path/To/File: Found 2 duplicates - deleting identical copies
```

**Comment by animosity22** [comment]
Can you please run the command with -vv and share the log? As a side note, checkers 24 makes thing slower as Google only allows about 10 transactions per second so if you use defaults, it'd go smooother.

**Comment by tcf909** [comment]
Logs attached.

I ran the command against the RawRemote (no Crypt). The result is the same against the Crypt remote.

[rclone-dedupe-logs.txt](https://github.com/rclone/rclone/files/4748552/rclone-dedupe-logs.txt)

**Comment by animosity22** [comment]
Those crypt entries make it tough!

Can we pick on directory to work on as that'll help narrow down some more logging as I'm going to assume if you run again, you see the same duplicates, right?

Let's pick this folder:

```
2020/06/08 20:48:48 NOTICE: Vault/jhjup0smhgu0n4splhvcoqj33c/6ur6j0s6gj2sn4u9ahnbpunoag/i24313maj3au9897v9e4992ja77hpcl4pe33vc8ob78gtfajmpvif416shseu0f324ej0dlhakf5t6835g3vbjf7hkfo2c7u2hb1v68/i81u8d3b1tqj3o3koq5aq5udqn29da96u0ph46ojbrgi1jl13ga3kof076d5omhs9oa49qpg2dn9pfu51fa8pn9nuvgl2rvd26btbs8: Found 2 duplicates - deleting identical copies
```

Check my eyes and I think this is the directory:

```
Vault/jhjup0smhgu0n4splhvcoqj33c/6ur6j0s6gj2sn4u9ahnbpunoag/i24313maj3au9897v9e4992ja77hpcl4pe33vc8ob78gtfajmpvif416shseu0f324ej0dlhakf5t6835g3vbjf7hkfo2c7u2hb1v68/
```

and the file is:

```
i81u8d3b1tqj3o3koq5aq5udqn29da96u0ph46ojbrgi1jl13ga3kof076d5omhs9oa49qpg2dn9pfu51fa8pn9nuvgl2rvd26btbs8
```

If you run a rclone ls on that with -vv, we should see the duplicate:

```
rlcone ls -vv RawRemote:Vault/jhjup0smhgu0n4splhvcoqj33c/6ur6j0s6gj2sn4u9ahnbpunoag/i24313maj3au9897v9e4992ja77hpcl4pe33vc8ob78gtfajmpvif416shseu0f324ej0dlhakf5t6835g3vbjf7hkfo2c7u2hb1v68
```

Let's start there.

**Comment by tcf909** [comment]
```
2020/06/08 21:13:00 DEBUG : Using config file from "/etc/rclone.conf"
    11035 i81u8d3b1tqj3o3koq5aq5udqn29da96u0ph46ojbrgi1jl13ga3kof076d5omhs9oa49qpg2dn9pfu51fa8pn9nuvgl2rvd26btbs8
    11035 i81u8d3b1tqj3o3koq5aq5udqn29da96u0ph46ojbrgi1jl13ga3kof076d5omhs9oa49qpg2dn9pfu51fa8pn9nuvgl2rvd26btbs8
       78 l9vva9it6pgvvin0m5kv4peva4
       78 l9vva9it6pgvvin0m5kv4peva4
1707505368 fevms5db6emag0n3ja0i7am2qm7gcaa5sgbk7peq43klpcglvl0a7c9d0im04jp61e6juli8mlhf0iqv9u7o4c5k9horpkfsl169hm0
2020/06/08 21:13:02 DEBUG : 4 go routines active
```

**Comment by tcf909** [comment]
`
# rclone dedupe -vv --dedupe-mode skip Raw:Vault/jhjup0smhgu0n4splhvcoqj33c/6ur6j0s6gj2sn4u9ahnbpunoag/i24313maj3au9897v9e4992ja77hpcl4pe33vc8ob78gtfajmpvif416shseu0f324ej0dlhakf5t6835g3vbjf7hkfo2c7u2hb1v68
2020/06/08 21:14:32 DEBUG : rclone: Version "v1.52.0" starting with parameters ["rclone" "dedupe" "-vv" "--dedupe-mode" "skip" "Raw:Vault/jhjup0smhgu0n4splhvcoqj33c/6ur6j0s6gj2sn4u9ahnbpunoag/i24313maj3au9897v9e4992ja77hpcl4pe33vc8ob78gtfajmpvif416shseu0f324ej0dlhakf5t6835g3vbjf7hkfo2c7u2hb1v68"]
2020/06/08 21:14:32 DEBUG : Using config file from "/etc/rclone.conf"
2020/06/08 21:14:34 INFO  : Google drive root 'Vault/jhjup0smhgu0n4splhvcoqj33c/6ur6j0s6gj2sn4u9ahnbpunoag/i24313maj3au9897v9e4992ja77hpcl4pe33vc8ob78gtfajmpvif416shseu0f324ej0dlhakf5t6835g3vbjf7hkfo2c7u2hb1v68': Looking for duplicates using skip mode.
2020/06/08 21:14:36 NOTICE: i81u8d3b1tqj3o3koq5aq5udqn29da96u0ph46ojbrgi1jl13ga3kof076d5omhs9oa49qpg2dn9pfu51fa8pn9nuvgl2rvd26btbs8: Found 2 duplicates - deleting identical copies
2020/06/08 21:14:36 NOTICE: l9vva9it6pgvvin0m5kv4peva4: Found 2 duplicates - deleting identical copies
2020/06/08 21:14:36 DEBUG : 4 go routines active
root@nas:/srv/Media# rclone dedupe -vv --dedupe-mode skip Raw:Vault/jhjup0smhgu0n4splhvcoqj33c/6ur6j0s6gj2sn4u9ahnbpunoag/i24313maj3au9897v9e4992ja77hpcl4pe33vc8ob78gtfajmpvif416shseu0f324ej0dlhakf5t6835g3vbjf7hkfo2c7u2hb1v68
2020/06/08 21:14:40 DEBUG : rclone: Version "v1.52.0" starting with parameters ["rclone" "dedupe" "-vv" "--dedupe-mode" "skip" "Raw:Vault/jhjup0smhgu0n4splhvcoqj33c/6ur6j0s6gj2sn4u9ahnbpunoag/i24313maj3au9897v9e4992ja77hpcl4pe33vc8ob78gtfajmpvif416shseu0f324ej0dlhakf5t6835g3vbjf7hkfo2c7u2hb1v68"]
2020/06/08 21:14:40 DEBUG : Using config file from "/etc/rclone.conf"
2020/06/08 21:14:42 INFO  : Google drive root 'Vault/jhjup0smhgu0n4splhvcoqj33c/6ur6j0s6gj2sn4u9ahnbpunoag/i24313maj3au9897v9e4992ja77hpcl4pe33vc8ob78gtfajmpvif416shseu0f324ej0dlhakf5t6835g3vbjf7hkfo2c7u2hb1v68': Looking for duplicates using skip mode.
2020/06/08 21:14:42 NOTICE: l9vva9it6pgvvin0m5kv4peva4: Found 2 duplicates - deleting identical copies
2020/06/08 21:14:42 NOTICE: i81u8d3b1tqj3o3koq5aq5udqn29da96u0ph46ojbrgi1jl13ga3kof076d5omhs9oa49qpg2dn9pfu51fa8pn9nuvgl2rvd26btbs8: Found 2 duplicates - deleting identical copies
2020/06/08 21:14:42 DEBUG : 4 go routines active
`

**Comment by tcf909** [comment]
It looks like the md5 is different on them.

```
2020/06/08 21:16:15 NOTICE: i81u8d3b1tqj3o3koq5aq5udqn29da96u0ph46ojbrgi1jl13ga3kof076d5omhs9oa49qpg2dn9pfu51fa8pn9nuvgl2rvd26btbs8: Found 2 duplicates - deleting identical copies
i81u8d3b1tqj3o3koq5aq5udqn29da96u0ph46ojbrgi1jl13ga3kof076d5omhs9oa49qpg2dn9pfu51fa8pn9nuvgl2rvd26btbs8: 2 duplicates remain
  1:        11035 bytes, 2017-10-04 01:23:02.752000000, MD5 e28c0010cfe694ee7aeb1a2245b98696
  2:        11035 bytes, 2017-10-04 01:23:02.752000000, MD5 f885603512b798f0d922ef738bab8260
```

**Comment by animosity22** [comment]
Yep, any crypt file is going to have a different MD5 on it.

Can you run the dedupe and add in --dump headers,responses

and log that. That should tell us what we need.

**Comment by tcf909** [comment]
[rclone-logs.txt](https://github.com/rclone/rclone/files/4748683/rclone-logs.txt)

**Comment by tcf909** [comment]
[rclone-logs.txt](https://github.com/rclone/rclone/files/4748694/rclone-logs.txt)

**Comment by animosity22** [comment]
So the piece I'm not seeing in the logs is any delete call being made. It lists them out as I can see the GET for the directory and I see two files in there that are the duplicates.

I definitely think we have all the detail to have @ncw take a peek.

For fun, are you able to test with --newest since they are identical anyway from my understanding? If that doesn't work, can you tryin --interactive and see if picking the one to delete works?

**Comment by tcf909** [comment]
I'm not really comfortable making changes to the raw remote (not sure what file I'm putting at risk) and don't want to have to manually try to translate names to files.

I did run dedupe interactive on the Crypt remote and saw this:

```
  1:   3927300780 bytes, 2017-09-24 06:41:25.868000000, None          hash type not supported
  2:   3927300780 bytes, 2017-09-24 06:41:25.868000000, None          hash type not supported
s/k/r> k
Enter the number of the file to keep> 1
2020/06/08 21:37:40 INFO  : (FILENAME REMOVED): Deleted
2020/06/08 21:37:40 NOTICE: (FILENAME REMOVED): Deleted 1 extra copies
```
When I ran dedupe against the file again, it still prompted me about the duplicate.

**Comment by animosity22** [comment]
You can use crypt decode as well albeit that's a pain so understandable.

So when you deleted the 1 copy and it said it deleted it, when you went back it was still there  in that interactive session?

**Comment by tcf909** [comment]
> So when you deleted the 1 copy and it said it deleted it, when you went back it was still there in that interactive session?

Correct.

**Comment by tcf909** [comment]
Here is the dump data after a file is selected to be deleted:

```
Enter the number of the file to keep> 1
2020/06/08 21:47:20 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2020/06/08 21:47:20 DEBUG : HTTP REQUEST (req 0xc0001d6d00)
2020/06/08 21:47:20 DEBUG : PATCH /drive/v3/files/0B4XodAnNYtzRb1hPdkF2dDVrM0k?alt=json&fields=&prettyPrint=false&supportsAllDrives=true HTTP/1.1
Host: www.googleapis.com
User-Agent: rclone/v1.52.0
Content-Length: 17
Authorization: XXXX
Content-Type: application/json
X-Goog-Api-Client: gl-go/1.14.3 gdcl/20200410
Accept-Encoding: gzip

2020/06/08 21:47:20 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2020/06/08 21:47:21 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2020/06/08 21:47:21 DEBUG : HTTP RESPONSE (req 0xc0001d6d00)
2020/06/08 21:47:21 DEBUG : HTTP/1.1 200 OK
Transfer-Encoding: chunked
Alt-Svc: h3-27=":443"; ma=2592000,h3-25=":443"; ma=2592000,h3-T050=":443"; ma=2592000,h3-Q050=":443"; ma=2592000,h3-Q049=":443"; ma=2592000,h3-Q048=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000,quic=":443"; ma=2592000; v="46,43"
Cache-Control: no-cache, no-store, max-age=0, must-revalidate
Content-Security-Policy: frame-ancestors 'self'
Content-Type: application/json; charset=UTF-8
Date: Mon, 08 Jun 2020 21:47:21 GMT
Expires: Mon, 01 Jan 1990 00:00:00 GMT
Pragma: no-cache
Server: GSE
Vary: Origin
Vary: X-Origin
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 1; mode=block

2
{}
0

2020/06/08 21:47:21 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
```

**Comment by tcf909** [comment]
Interestingly enough. Now the file is no longer duplicated. 

(ran the same command 3 times, the 3rd time having the --dump args)

**Comment by animosity22** [comment]
So this worked for me on my crypt mount:

```
felix@gemini:~$ rclone dedupe gcrypt:test -vv
2020/06/08 17:50:00 DEBUG : rclone: Version "v1.52.0" starting with parameters ["rclone" "dedupe" "gcrypt:test" "-vv"]
2020/06/08 17:50:00 DEBUG : Using config file from "/opt/rclone/rclone.conf"
2020/06/08 17:50:01 INFO  : Encrypted drive 'gcrypt:test': Looking for duplicates using interactive mode.
2020/06/08 17:50:01 NOTICE: test.mkv: Found 2 duplicates - deleting identical copies
test.mkv: 2 duplicates remain
  1:   1504953150 bytes, 2020-06-08 17:48:38.572000000, None          hash type not supported
  2:   1504953150 bytes, 2020-06-08 06:17:00.527000000, None          hash type not supported
s) Skip and do nothing
k) Keep just one (choose which in next step)
r) Rename all to be different (by changing file.jpg to file-1.jpg)
s/k/r> k
Enter the number of the file to keep> 1
2020/06/08 17:50:12 INFO  : test.mkv: Deleted
2020/06/08 17:50:12 NOTICE: test.mkv: Deleted 1 extra copies
2020/06/08 17:50:12 DEBUG : 6 go routines active
felix@gemini:~$
```

Note the deleted 1 extra copies line.

and this does not work:

```
felix@gemini:~$ rclone ls gcrypt:test
1504953150 test.mkv
1504953150 test.mkv
felix@gemini:~$ rclone dedupe gcrypt:test -vv --dedupe-mode skip
2020/06/08 17:51:21 DEBUG : rclone: Version "v1.52.0" starting with parameters ["rclone" "dedupe" "gcrypt:test" "-vv" "--dedupe-mode" "skip"]
2020/06/08 17:51:21 DEBUG : Using config file from "/opt/rclone/rclone.conf"
2020/06/08 17:51:22 INFO  : Encrypted drive 'gcrypt:test': Looking for duplicates using skip mode.
2020/06/08 17:51:23 NOTICE: test.mkv: Found 2 duplicates - deleting identical copies
2020/06/08 17:51:23 DEBUG : 6 go routines active
felix@gemini:~$
```

So it's easy to recreate to as I just made a copy of a crypt file and renamed it the same in the drive UI.

**Comment by animosity22** [comment]
Newest works for me as well:

```
felix@gemini:~$ rclone dedupe gcrypt:test -vv --dedupe-mode newest
2020/06/08 17:52:36 DEBUG : rclone: Version "v1.52.0" starting with parameters ["rclone" "dedupe" "gcrypt:test" "-vv" "--dedupe-mode" "newest"]
2020/06/08 17:52:36 DEBUG : Using config file from "/opt/rclone/rclone.conf"
2020/06/08 17:52:37 INFO  : Encrypted drive 'gcrypt:test': Looking for duplicates using newest mode.
2020/06/08 17:52:37 NOTICE: test.mkv: Found 2 duplicates - deleting identical copies
2020/06/08 17:52:38 INFO  : test.mkv: Deleted
2020/06/08 17:52:38 NOTICE: test.mkv: Deleted 1 extra copies
2020/06/08 17:52:38 DEBUG : 6 go routines active
```

**Comment by tcf909** [comment]
OK. Glad you are able to recreate.

**Comment by animosity22** [comment]
The better we can frame an issue, the less time Nick has to spend doing that and he can focus on the code.

One of these days I'll try to learn more programming :)

**Comment by ncw** [comment]
I think this is a bug in the logging...

According to the docs

> --dedupe-mode skip - removes identical files then skips anything left.

And I think rclone **is** doing that. The files in questions are different because they have different MD5 sums so rclone didn't delete them.

However it did print

> 2020/06/08 19:26:07 NOTICE: Path/To/File: Found 2 duplicates - deleting identical copies

Which is a very misleading log!

What it actually means according to the code is

- Rclone found 2 files with duplicate file names
- It is now going to see if there are any identical files
- and **if** there are any it will remove them

However I think we've all been reading the log as meaning

- Rclone found 2 files which are identical and will delete them

If rclone does find identical files it will print this log

>Deleting %d/%d identical duplicates (%v %q)

But as @animosity22 pointed out,, duplicate files on crypt don't have the same md5sum (because of the encryption is different each time).

I think we can fix the confusion if I make the first log say

> 2020/06/08 19:26:07 NOTICE: Path/To/File: Found 2 files with duplicate names

And add an addition log when using "skip" (there are no logs for it at the moment) saying

> 2020/06/08 19:26:07 NOTICE: Path/To/File: Skipping deletion of 2 files with duplicate names

Would that work?

**Comment by animosity22** [comment]
Ok, that makes sense to me.

So it basically fails because it is on a crypt and it can't find the md5sum to say they are name if I'm getting that.

As an example:

```
felix@gemini:/GD/test$ rclone dedupe gcrypt:test -vv --dedupe-mode skip
2020/06/14 07:49:35 DEBUG : rclone: Version "v1.52.1" starting with parameters ["rclone" "dedupe" "gcrypt:test" "-vv" "--dedupe-mode" "skip"]
2020/06/14 07:49:35 DEBUG : Using config file from "/opt/rclone/rclone.conf"
2020/06/14 07:49:36 INFO  : Encrypted drive 'gcrypt:test': Looking for duplicates using skip mode.
2020/06/14 07:49:36 NOTICE: hosts: Found 2 duplicates - deleting identical copies
2020/06/14 07:49:36 DEBUG : 6 go routines active
felix@gemini:/GD/test$ rclone dedupe gcrypt:test -vv --dedupe-mode skip
2020/06/14 07:50:39 DEBUG : rclone: Version "v1.52.1" starting with parameters ["rclone" "dedupe" "gcrypt:test" "-vv" "--dedupe-mode" "skip"]
2020/06/14 07:50:39 DEBUG : Using config file from "/opt/rclone/rclone.conf"
2020/06/14 07:50:40 INFO  : Encrypted drive 'gcrypt:test': Looking for duplicates using skip mode.
2020/06/14 07:50:40 NOTICE: hosts: Found 2 duplicates - deleting identical copies
2020/06/14 07:50:40 DEBUG : 6 go routines active
```

and running that on the non encrypted:

```
felix@gemini:/GD/test$ rclone dedupe GD:crypt/191qpocr6dhlo35vn2gk2h8gf4 -vv --dedupe-mode skip
2020/06/14 07:51:08 DEBUG : rclone: Version "v1.52.1" starting with parameters ["rclone" "dedupe" "GD:crypt/191qpocr6dhlo35vn2gk2h8gf4" "-vv" "--dedupe-mode" "skip"]
2020/06/14 07:51:08 DEBUG : Using config file from "/opt/rclone/rclone.conf"
2020/06/14 07:51:08 INFO  : Google drive root 'crypt/191qpocr6dhlo35vn2gk2h8gf4': Looking for duplicates using skip mode.
2020/06/14 07:51:09 NOTICE: rs88l6p51j4kp0g9if3j294ts0: Found 2 duplicates - deleting identical copies
2020/06/14 07:51:09 NOTICE: rs88l6p51j4kp0g9if3j294ts0: Deleting 1/2 identical duplicates (MD5 "bfbf843b1e40dd3ebe4a7c0c33fa396a")
2020/06/14 07:51:09 INFO  : rs88l6p51j4kp0g9if3j294ts0: Deleted
2020/06/14 07:51:09 NOTICE: rs88l6p51j4kp0g9if3j294ts0: All duplicates removed
2020/06/14 07:51:09 DEBUG : 4 go routines active
```

**Comment by tcf909** [comment]
definitely fixing the logs makes sense. it should be clear that two files have different md5s and that nothing was done.

probably separate issue worthy, but wouldn't it make sense to add two flags to resolve the intent of the behavior:

1) --size-only (can apply to any remote but especially useful for crypt remotes)

2) --crypt-check (use crypt check function to check files)

I'm really reluctant to deal with the TB's of dupes interactively and without a size comparison option I'm afraid I'll delete a file that is truly unique with one of the dedupe modes.

thoughts?

**Comment by ncw** [comment]
I've merged the loging fix to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.53



> probably separate issue worthy, but wouldn't it make sense to add two flags to resolve the intent of the behavior:
> 
> 1. --size-only (can apply to any remote but especially useful for crypt remotes)

That is a good idea...

I had a go at implementing it here

https://beta.rclone.org/branch/v1.52.1-077-g388fa32f-fix-4321-dedupe-beta/ (uploaded in 15-30 mins)

Can you have a go?

> -crypt-check (use crypt check function to check files)

That would probably involve downloading the files to check their equality...

**Comment by tcf909** [comment]
> -crypt-check (use crypt check function to check files

has rclone looked at using any xattrs (local) or extended meta (remote)?

I get the sense that a design philosophy that exists is (change no data), but there is a lot of cool things that could be done with a little state.

Storing a common encrypted file hash for instant could allow you to handle files more like a normal file (sort of like what you did with vfs for mount).

**Comment by tcf909** [comment]
> Can you have a go?
Both the updated message and the --size-only flag work great.
Thanks Nick.

> That would probably involve downloading the files to check their equality...

Maybe --source-dir /local/cache/of/files ?

I still think the better overall solution would be to store an encrypted hash in an extended attribute. Crypt could then transparently provide the hash on call via reading the extended attribute, decrypting it, then providing the real hash. My Go skills are...non-existent these days or I would take a stab at it.

The attribute could be stored in most remotes I believe and then could use xattrs locally.

**Comment by animosity22** [comment]
There is already an  issue somewhere for a rewrite for crypt v2 as that would handle that item.

For now, that's great idea, but doesn't work with every remote and you have to think of Windows and Linux as rclone runs on many, many solutions.

**Comment by tcf909** [comment]
>For now, that's great idea, but doesn't work with every remote and you have to think of Windows and Linux as rclone runs on many, many solutions.

I'm pretty sure extended attributes on both windows and linux would be supported (as long as it was NTFS on windows).

**Comment by animosity22** [comment]
We're going a bit off tangent on the issue and I should find the v2 crypt item as that's more relevant.

I understand your point as the challenge is you making customizations per OS and it doesn't scale / maintain well. Existence is one thing and having to validate all the bits support it is another. Nick already has an item for crypt rewrite that's down the road so best to find that and discuss it there rather than this issue as the feedback is always weclome.

**Comment by animosity22** [comment]
Here you go:

https://github.com/rclone/rclone/issues/3667

Please add feedback and suggestions as I suggested so it gets in the right spot.

Thanks!

**Comment by ncw** [comment]
I've merged the dedupe --size-only fix to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.53
